File: app/utils/dawproject/track.ts
import { Channel } from "./channel";
import { ContentType } from "./contentType";
import { Lane } from "./lane";
import { registerLane } from "./registry/laneRegistry";
import { ITrack } from "./types";
import { Utility } from "./utility";

const trackFactory = (xmlObject: any): Track => {
  const instance = new Track();
  instance.fromXmlObject(xmlObject);
  return instance;
};

@registerLane("Track", trackFactory)
/** Represents a sequencer track.  */
export class Track extends Lane implements ITrack {
  /** Role of this track in timelines & arranger. Can be multiple (comma-separated). */
  contentType: ContentType[];
  /** If this track is loaded/active of not. */
  loaded?: boolean;
  /** Mixer channel used for the output of this track. */
  channel?: Channel;
  /** Child tracks, typically used to represent group/folder tracks with contentType="tracks". */
  tracks: Track[]; // Nested tracks

  constructor(
    contentType: ContentType[] = [],
    loaded?: boolean,
    channel?: Channel,
    tracks?: Track[],
    name?: string,
    color?: string,
    comment?: string
  ) {
    super(name, color, comment);
    this.contentType = contentType;
    this.loaded = loaded;
    this.channel = channel;
    this.tracks = tracks || [];
  }

  toXmlObject(): any {
    const obj: any = {
      Track: {
        ...super.toXmlObject(), // get attributes from Lane
      },
    };

    // add optional attributes
    // contentType is a list of strings, so we join them with commas
    Utility.addAttribute(obj.Track, "contentType", this, {
      adapter: (value: ContentType[]) => value.join(","),
    });
    Utility.addAttribute(obj.Track, "loaded", this);

    // append Channel as a nested XML element if present
    if (this.channel) {
      obj.Track.Channel = this.channel.toXmlObject().Channel;
    }

    // recursively add nested tracks
    if (this.tracks && this.tracks.length > 0) {
      obj.Track.Track = this.tracks.map(track => track.toXmlObject().Track);
    }

    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Lane

    // extract contentType text and split into a list
    Utility.populateAttribute<ContentType[]>(xmlObject, "contentType", this, {
      adapter: (value: string) => value.split(",") as ContentType[],
    });

    // parse the loaded attribute, converting 'true'/'false' to Boolean
    Utility.populateAttribute<boolean>(xmlObject, "loaded", this, {
      castTo: Boolean,
    });

    // initialize channel using Channel's fromXmlObject method if present
    if (xmlObject.Channel) {
      this.channel = new Channel().fromXmlObject(xmlObject.Channel);
    }

    // recursively parse nested Track elements
    const tracks: Track[] = [];
    if (xmlObject.Track) {
      const trackArray = Array.isArray(xmlObject.Track)
        ? xmlObject.Track
        : [xmlObject.Track];
      trackArray.forEach((trackObj: any) => {
        tracks.push(new Track().fromXmlObject(trackObj));
      });
    }
    this.tracks = tracks;

    return this;
  }
}

File: app/utils/dawproject_python/track.py
from lxml import etree as ET
from classes.lane import Lane
from classes.channel import Channel


class Track(Lane):
    def __init__(
        self,
        content_type=None,
        loaded=None,
        channel=None,
        tracks=None,
        name=None,
        color=None,
        comment=None,
    ):
        super().__init__(name, color, comment)
        self.content_type = content_type if content_type else []
        self.loaded = loaded
        self.channel = channel
        self.tracks = tracks if tracks else []

    def to_xml(self):
        # Inheriting to_xml from Lane and changing tag
        track_elem = super().to_xml()
        track_elem.tag = "Track"

        # Set content_type as an attribute
        if self.content_type:
            content_type_str = ",".join(str(ct.value) for ct in self.content_type)
            track_elem.set("contentType", content_type_str)

        # Set loaded as an attribute, converting to lowercase string for XML
        if self.loaded is not None:
            track_elem.set("loaded", str(self.loaded).lower())

        # Append Channel as a nested XML element if present
        if self.channel:
            channel_elem = ET.SubElement(track_elem, "Channel")
            channel_elem.extend(self.channel.to_xml().getchildren())

        # Recursively add nested tracks
        for track in self.tracks:
            track_elem.append(track.to_xml())

        return track_elem

    @classmethod
    def from_xml(cls, element):
        # Call the superclass method to initialize inherited attributes
        instance = super().from_xml(element)

        # Extract contentType text and split into a list
        content_type_elem = element.find("contentType")
        if content_type_elem is not None and content_type_elem.text:
            instance.content_type = content_type_elem.text.split(",")

        # Parse the loaded attribute, converting 'true'/'false' to Boolean
        loaded = element.get("loaded")
        instance.loaded = loaded.lower() == "true" if loaded else None

        # Initialize channel using Channel's from_xml method if present
        channel_elem = element.find("Channel")
        instance.channel = (
            Channel.from_xml(channel_elem) if channel_elem is not None else None
        )

        # Recursively parse nested Track elements
        tracks = []
        for track_elem in element.findall("Track"):
            tracks.append(Track.from_xml(track_elem))
        instance.tracks = tracks

        return instance
